driver/scsi_operation.c

```diff
@@ -30,55 +30,85 @@ UCHAR
 WnbdReadCapacity(_In_ PSCSI_DEVICE_INFORMATION Info,
                  _In_ PSCSI_REQUEST_BLOCK Srb,
                  _In_ PCDB Cdb,
-                 _In_ UINT64 BlockSize,
+                 _In_ UINT BlockSize,
                  _In_ UINT64 BlockCount)
 {
     WNBD_LOG_LOUD(": Enter");
     ASSERT(Srb);
     ASSERT(Cdb);
-    WNBD_LOG_INFO("Using BlockCount: %llu", BlockCount);
+    WNBD_LOG_INFO("Using BlockSize: %u, BlockCount: %llu", BlockSize, BlockCount);
 
     PVOID DataBuffer = SrbGetDataBuffer(Srb);
     ULONG DataTransferLength = SrbGetDataTransferLength(Srb);
-    UCHAR SrbStatus = SRB_STATUS_DATA_OVERRUN;
+    UCHAR SrbStatus = SRB_STATUS_INVALID_REQUEST;
 
-    if (0 == DataBuffer) {
+    if (0 == DataBuffer || 0 == BlockCount) {
+        SrbStatus = SRB_STATUS_INTERNAL_ERROR;
         goto Exit;
     }
+    UINT64 maxLba = BlockCount - 1;
 
     RtlZeroMemory(DataBuffer, DataTransferLength);
 
-    switch (Cdb->AsByte[0])
+    switch (Cdb->CDB10.OperationCode)
     {
     case SCSIOP_READ_CAPACITY:
         {
         if (sizeof(READ_CAPACITY_DATA) > DataTransferLength) {
             goto Exit;
         }
+
         PREAD_CAPACITY_DATA ReadCapacityData = DataBuffer;
-        REVERSE_BYTES_4(&ReadCapacityData->LogicalBlockAddress, &BlockCount);
+
+        if (maxLba >= MAXULONG) {
+            ReadCapacityData->LogicalBlockAddress = MAXULONG;
+        }
+        else {
+            REVERSE_BYTES_4(&ReadCapacityData->LogicalBlockAddress, &maxLba);
+        }
+
         REVERSE_BYTES_4(&ReadCapacityData->BytesPerBlock, &BlockSize);
         SrbSetDataTransferLength(Srb, sizeof(READ_CAPACITY_DATA));
         SrbStatus = SRB_STATUS_SUCCESS;
         }
         break;
     case SCSIOP_READ_CAPACITY16:
         {
-        if (sizeof(READ_CAPACITY16_DATA) > DataTransferLength) {
+        if (Cdb->READ_CAPACITY16.ServiceAction != SERVICE_ACTION_READ_CAPACITY16) {
+            goto Exit;
+        }
+
+        if (sizeof(READ_CAPACITY_DATA_EX) > DataTransferLength) {
             goto Exit;
         }
+
+        ULONG ReturnDataLength = 0;
         PREAD_CAPACITY16_DATA ReadCapacityData16 = DataBuffer;
-        REVERSE_BYTES_8(&ReadCapacityData16->LogicalBlockAddress, &BlockCount);
-        REVERSE_BYTES_8(&ReadCapacityData16->BytesPerBlock, &BlockSize);
-        if (CHECK_NBD_SEND_TRIM(Info->UserEntry->NbdFlags)) {
-            ReadCapacityData16->LBPME = 1;
+        REVERSE_BYTES_8(&ReadCapacityData16->LogicalBlockAddress, &maxLba);
+        REVERSE_BYTES_4(&ReadCapacityData16->BytesPerBlock, &BlockSize);
+
+        if (DataTransferLength >= (ULONG)FIELD_OFFSET(READ_CAPACITY16_DATA, Reserved3)) {
+            if (CHECK_NBD_SEND_TRIM(Info->UserEntry->NbdFlags)) {
```

Comment:
This will have to be replaced with the following, as per recent changes:

`if (Info->UserEntry->Properties.Flags.UnmapSupported) {`